### ISSUE TITLE
multiprocessing spawn in python 3 does not work on macOS

### ISSUE THREAD

**Issue by simongus** [pyinstaller/pyinstaller#2322]
I'm not able to get multiprocessing to work in spawn mode on macOS  when using pyinstaller. The following script runs fine in the interpreter, but when using pyinstaller it enters an infinite recursive loop that keeps starting new processes.

I'm using Python 3 from Anaconda (Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:52:12) ), on macOS 10.12.2, with the development version of pyinstaller (3.3.dev0+483c819).

```python
import sys
from multiprocessing import freeze_support, set_start_method, Process

def test():
    print('In subprocess')

if __name__ == '__main__':
    print(sys.argv,)
    freeze_support()
    set_start_method('spawn')

    print('In main')
    proc = Process(target=test)
    proc.start()
    proc.join()
```

**Comment by xoviat** [comment]
The multithreading tests are currently failing on linux for everything except Python34. My guess is that this is somehow related. I would help but there's not much that I can do to debug this because I'm on windows.

**Comment by mkassner** [comment]
I can confirm the same using homebrew installed python3.6 and the current (11f2e7b) branch of pyinstaller . 

This is a blocker for us. Is there anything we can do to help this issue get resolved?

below the output of `print(sys.argv,)`

```shell
['./dist/Pupil Capture.app/Contents/MacOS/pupil_capture']
['/Users/mkassner/Pupil/pupil_code/deployment/deploy_capture/dist/Pupil Capture.app/Contents/MacOS/pupil_capture', '-B', '-s', '-S', '-E', '-c', 'from multiprocessing.semaphore_tracker import main;main(7)']
['/Users/mkassner/Pupil/pupil_code/deployment/deploy_capture/dist/Pupil Capture.app/Contents/MacOS/pupil_capture', '--multiprocessing-fork', 'tracker_fd=8', 'pipe_handle=57']
['/Users/mkassner/Pupil/pupil_code/deployment/deploy_capture/dist/Pupil Capture.app/Contents/MacOS/pupil_capture', '-B', '-s', '-S', '-E', '-c', 'from multiprocessing.semaphore_tracker import main;main(8)']
['/Users/mkassner/Pupil/pupil_code/deployment/deploy_capture/dist/Pupil Capture.app/Contents/MacOS/pupil_capture', '--multiprocessing-fork', 'tracker_fd=9', 'pipe_handle=63']
['/Users/mkassner/Pupil/pupil_code/deployment/deploy_capture/dist/Pupil Capture.app/Contents/MacOS/pupil_capture', '-B', '-s', '-S', '-E', '-c', 'from multiprocessing.semaphore_tracker import main;main(9)']
['/Users/mkassner/Pupil/pupil_code/deployment/deploy_capture/dist/Pupil Capture.app/Contents/MacOS/pupil_capture', '-B', '-s', '-S', '-E', '-c', 'from multiprocessing.semaphore_tracker import main;main(7)']
['/Users/mkassner/Pupil/pupil_code/deployment/deploy_capture/dist/Pupil Capture.app/Contents/MacOS/pupil_capture', '--multiprocessing-fork', 'tracker_fd=10', 'pipe_handle=66']
['/Users/mkassner/Pupil/pupil_code/deployment/deploy_capture/dist/Pupil Capture.app/Contents/MacOS/pupil_capture', '--multiprocessing-fork', 'tracker_fd=9', 'pipe_handle=63']
['/Users/mkassner/Pupil/pupil_code/deployment/deploy_capture/dist/Pupil Capture.app/Contents/MacOS/pupil_capture', '-B', '-s', '-S', '-E', '-c', 'from multiprocessing.semaphore_tracker import main;main(8)']
['/Users/mkassner/Pupil/pupil_code/deployment/deploy_capture/dist/Pupil Capture.app/Contents/MacOS/pupil_capture', '-B', '-s', '-S', '-E', '-c', 'from multiprocessing.semaphore_tracker import main;main(7)']
```

**Comment by xoviat** [comment]
We probably should get the output from the debug bootloaders. Rename the debug bootloaders [here][1] to the regular bootloaders and then capture the stderr from the processes. This is [what we're looking for][2].

[1]: https://github.com/pyinstaller/pyinstaller/tree/11f2e7b78bd2f28706bff7cefb4e76c7ea0d7b4c/PyInstaller/bootloader/Darwin-64bit
[2]: https://travis-ci.org/xoviat/pyinstaller/jobs/198261598#L1633

**Comment by htgoebel** [comment]
No need to rename the bootloader. Just run PyInstalller with `--debug` -- the the fine manual.

**Comment by htgoebel** [comment]
@simongus Did you follow https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Multiprocessing?

**Comment by simongus** [comment]
I did.  However, those recipes only apply to Windows, not macOS.

**Comment by mkassner** [comment]
Hi,

here the output from running a bundle with --debug: https://gist.github.com/mkassner/e56c17e7c1c65e8a9d03ef411f7499b3

let me know what else I can supply.

**Comment by xoviat** [comment]
Just what I suspected actually `_MEIPASS2` is null on the subprocesses. Now why could that be?

**Comment by xoviat** [comment]
Actually not what I suspected. Let me see what this should look like...

**Comment by xoviat** [comment]
Should be [this](https://gist.github.com/xoviat/d3f612ae4b402addcbb3f8a047b65745)

**Comment by mkassner** [comment]
I m not sure I follow. Did I not provide the correct output or is my fist indicative of something going wrong. 

To reiterate if I use start method: `fork` it works. But I need to use `spawn`. This triggers an endless loop.

**Comment by xoviat** [comment]
> I m not sure I follow. Did I not provide the correct output or is my fist indicative of something going wrong.

No. The information that I provided was for my own reference.

**Comment by xoviat** [comment]
It appears that my suspicions about MEIPASS2 were originally correct. Try putting this before all  other code:

```
import os
import sys

# Module multiprocessing is organized differently in Python 3.4+
# Python 3.4+
import multiprocessing.popen_spawn_posix as forking

# First define a modified version of Popen.
class _Popen(forking.Popen):
    def __init__(self, *args, **kw):
        if hasattr(sys, 'frozen'):
            # We have to set original _MEIPASS2 value from sys._MEIPASS
            # to get --onefile mode working.
            os.putenv('_MEIPASS2', sys._MEIPASS)
        try:
            super(_Popen, self).__init__(*args, **kw)
        finally:
            if hasattr(sys, 'frozen'):
                # On some platforms (e.g. AIX) 'os.unsetenv()' is not
                # available. In those cases we cannot delete the variable
                # but only set it to the empty string. The bootloader
                # can handle this case.
                if hasattr(os, 'unsetenv'):
                    os.unsetenv('_MEIPASS2')
                else:
                    os.putenv('_MEIPASS2', '')

# Second override 'Popen' class with our modified version.
forking.Popen = _Popen
```

**Comment by mkassner** [comment]
I added what you asked for and can confirm that the modified  Popen is being used but it did not solve the endless startup loop problem.

Here to log: https://gist.github.com/mkassner/029472f96b800ba0e42acfb1ff7208fe

**Comment by xoviat** [comment]
Keeping what you have, place this before all other code:


```python
def get_command_line(**kwds):
    '''
    Returns prefix of command line used for spawning a child process
    '''
    return ([sys.executable, '--multiprocessing-fork'] +
            ['%s=%r' % item for item in kwds.items()])

import multiprocessing.spawn
multiprocessing.spawn.get_command_line = get_command_line
```

If this doesn't work, then I'll probably write a test for this and then I can probe it on travis...

**Comment by mkassner** [comment]
still not working I m afraid. Here the log:

https://gist.github.com/mkassner/f76baf6b252c47dca08452628f56c825

**Comment by htgoebel** [comment]
this gist does not match your small example about: The is zmq mentioned, while your example does not use it. Please test with your minimal example to avoid other problems interfering with this one.

**Comment by mkassner** [comment]
sorry about that.

I have rerun the correct script see below:
```python
'''
(*)~---------------------------------------------------------------------------
Pupil - eye tracking platform
Copyright (C) 2012-2017  Pupil Labs

Distributed under the terms of the GNU
Lesser General Public License (LGPL v3.0).
See COPYING and COPYING.LESSER for license details.
---------------------------------------------------------------------------~(*)
'''

import sys, os

def get_command_line(**kwds):
    print('DEBUG: get_command_line called')
    '''
    Returns prefix of command line used for spawning a child process
    '''
    return ([sys.executable, '--multiprocessing-fork'] +
            ['%s=%r' % item for item in kwds.items()])

import multiprocessing.spawn
multiprocessing.spawn.get_command_line = get_command_line


import os
import sys

# Module multiprocessing is organized differently in Python 3.4+
# Python 3.4+
import multiprocessing.popen_spawn_posix as forking

# First define a modified version of Popen.
class _Popen(forking.Popen):
    def __init__(self, *args, **kw):
        if hasattr(sys, 'frozen'):
            # We have to set original _MEIPASS2 value from sys._MEIPASS
            # to get --onefile mode working.
            os.putenv('_MEIPASS2', sys._MEIPASS)
        try:
            super(_Popen, self).__init__(*args, **kw)
        finally:
            if hasattr(sys, 'frozen'):
                # On some platforms (e.g. AIX) 'os.unsetenv()' is not
                # available. In those cases we cannot delete the variable
                # but only set it to the empty string. The bootloader
                # can handle this case.
                if hasattr(os, 'unsetenv'):
                    os.unsetenv('_MEIPASS2')
                else:
                    os.putenv('_MEIPASS2', '')

# Second override 'Popen' class with our modified version.
forking.Popen = _Popen

import sys
from multiprocessing import freeze_support, set_start_method, Process

def test():
    print('In subprocess')

if __name__ == '__main__':
    print(sys.argv,)
    freeze_support()
    set_start_method('spawn')

    print('In main')
    proc = Process(target=test)
    proc.start()
    proc.join()
```

the problem behaviour is the same.

the log output is here: https://gist.github.com/mkassner/7d1dda745ab33180510a98af34d7862f

**Comment by mkassner** [comment]
Please let me know if there is anything else I can do to help with this issue. 

pyinstaller is a great tool and we use it for deployment, unfortunately we need to use `spawn` on MacOS. We used billiard and  python2 in the past successfully with pyinstaller. Now with the migration of the project to python3 we are no longer able to bundle for mac...

If you give me pointers I m happy to do more tests/investigations!

**Comment by xoviat** [comment]
@mkassner If you're wondering why I haven't asked for additional information, it's because this is a harder problem and I need to figure out what might be happening before I proceed.

**Comment by mkassner** [comment]
@xoviat  fully understood. I know what you mean! We would be happy to put a bounty on this if we can expedite the process! (But I dont know if this is something the pyinstaller team is ok with...)

**Comment by xoviat** [comment]
Alright: I've put something together that will (finally, hopefully) capture enough output to understand what is happening. As always, simply put this before all else and then post the output:

```python
import os
import sys
from contextlib import suppress

pid = os.getpid()

def trace_calls(frame, event, arg):
    if event != 'call':
        return
    co = frame.f_code
    func_name = co.co_name
    if func_name == 'write':
        # Ignore write() calls from print statements
        return
    elif event == 'return':
        print('[{}] {} => {}'.format(pid, func_name, arg))
        return
    line_no = frame.f_lineno
    filename = co.co_filename
    
    if 'multiprocessing' not in filename:
        return
    
    #with suppress(Exception):
    print('[{}] Call to {} on line {} of {}'.format(pid, func_name, line_no, filename))
    
    for i in range(frame.f_code.co_argcount):
    #    with suppress(Exception):
        name = frame.f_code.co_varnames[i]
        if name in frame.f_locals:
            value = frame.f_locals[name]
        elif name in frame.f_globals:
            value = frame.f_globals[name]
        else:
            value = ''
            
        if type(value) is not str:
            with suppress(Exception):
                value = repr(value)
        
        if type(value) is not str:
            with suppress(Exception):
                value = str(value)
        
        if type(value) is not str:
            value = ''
        
        print("[{}]    Argument {} is {}".format(pid, name, value))
   
    
sys.settrace(trace_calls)
```

**Comment by mkassner** [comment]
I bundled and ran the below code:

```python
import os
import sys
from contextlib import suppress

pid = os.getpid()

def trace_calls(frame, event, arg):
    if event != 'call':
        return
    co = frame.f_code
    func_name = co.co_name
    if func_name == 'write':
        # Ignore write() calls from print statements
        return
    elif event == 'return':
        print('[{}] {} => {}'.format(pid, func_name, arg))
        return
    line_no = frame.f_lineno
    filename = co.co_filename

    if 'multiprocessing' not in filename:
        return

    #with suppress(Exception):
    print('[{}] Call to {} on line {} of {}'.format(pid, func_name, line_no, filename))

    for i in range(frame.f_code.co_argcount):
    #    with suppress(Exception):
        name = frame.f_code.co_varnames[i]
        if name in frame.f_locals:
            value = frame.f_locals[name]
        elif name in frame.f_globals:
            value = frame.f_globals[name]
        else:
            value = ''

        if type(value) is not str:
            with suppress(Exception):
                value = repr(value)

        if type(value) is not str:
            with suppress(Exception):
                value = str(value)

        if type(value) is not str:
            value = ''

        print("[{}]    Argument {} is {}".format(pid, name, value))


sys.settrace(trace_calls)




import sys, os

def get_command_line(**kwds):
    print('DEBUG: get_command_line called')
    '''
    Returns prefix of command line used for spawning a child process
    '''
    return ([sys.executable, '--multiprocessing-fork'] +
            ['%s=%r' % item for item in kwds.items()])

import multiprocessing.spawn
multiprocessing.spawn.get_command_line = get_command_line


import os
import sys

# Module multiprocessing is organized differently in Python 3.4+
# Python 3.4+
import multiprocessing.popen_spawn_posix as forking

# First define a modified version of Popen.
class _Popen(forking.Popen):
    def __init__(self, *args, **kw):
        if hasattr(sys, 'frozen'):
            # We have to set original _MEIPASS2 value from sys._MEIPASS
            # to get --onefile mode working.
            os.putenv('_MEIPASS2', sys._MEIPASS)
        try:
            super(_Popen, self).__init__(*args, **kw)
        finally:
            if hasattr(sys, 'frozen'):
                # On some platforms (e.g. AIX) 'os.unsetenv()' is not
                # available. In those cases we cannot delete the variable
                # but only set it to the empty string. The bootloader
                # can handle this case.
                if hasattr(os, 'unsetenv'):
                    os.unsetenv('_MEIPASS2')
                else:
                    os.putenv('_MEIPASS2', '')

# Second override 'Popen' class with our modified version.
forking.Popen = _Popen

import sys
from multiprocessing import freeze_support, set_start_method, Process

def test():
    print('In subprocess')

if __name__ == '__main__':
    print(sys.argv,)
    freeze_support()
    set_start_method('spawn')

    print('In main')
    proc = Process(target=test)
    proc.start()
    proc.join()
```

This is the output: https://gist.github.com/mkassner/b4cd2403dece04d926181f593736b5ce

I hope it gives insight!

**Comment by xoviat** [comment]
Alright, let's try this:

```python
import os
import sys
from contextlib import suppress

pid = os.getpid()

def trace_calls(frame, event, arg):
    if event != 'call':
        return
    co = frame.f_code
    func_name = co.co_name
    if func_name == 'write':
        # Ignore write() calls from print statements
        return
    elif event == 'return':
        print('[{}] {} => {}'.format(pid, func_name, arg))
        return
    line_no = frame.f_lineno
    filename = co.co_filename

    c_frame = frame
    c_filename = filename
    c_line_no = line_no

    with suppress(Exception):
        for i in range(10):
            if not (c_line_no == line_no and c_filename == filename):
                break

            c_frame = frame.f_back
            c_line_no = c_frame.f_lineno
            c_filename = c_frame.f_code.co_filename
        else:
            c_line_no = 'UNKNOWN'
            c_filename = 'UNKNOWN'

    if 'multiprocessing' not in filename:
        return

    #with suppress(Exception):
    print('[{}] Call to {} on line {} of {} from {} of {}'.format(pid, func_name, line_no, filename, c_line_no, c_filename))

    for i in range(frame.f_code.co_argcount):
    #    with suppress(Exception):
        name = frame.f_code.co_varnames[i]
        if name in frame.f_locals:
            value = frame.f_locals[name]
        elif name in frame.f_globals:
            value = frame.f_globals[name]
        else:
            value = ''

        if type(value) is not str:
            with suppress(Exception):
                value = repr(value)

        if type(value) is not str:
            with suppress(Exception):
                value = str(value)

        if type(value) is not str:
            value = ''

        print("[{}]    Argument {} is {}".format(pid, name, value))


sys.settrace(trace_calls)

import sys, os

def get_command_line(**kwds):
    print('DEBUG: get_command_line called')
    '''
    Returns prefix of command line used for spawning a child process
    '''
    return ([sys.executable, '--multiprocessing-fork'] +
            ['%s=%r' % item for item in kwds.items()])

from multiprocessing.spawn import os, process, reduction, prepare

def _main(fd):
    with os.fdopen(fd, 'rb', closefd=True) as from_parent:
        process.current_process()._inheriting = True
        try:
            preparation_data = reduction.pickle.load(from_parent)
            prepare(preparation_data)
            self = reduction.pickle.load(from_parent)
        except Exception as e:
            
        finally:
            del process.current_process()._inheriting
    return self._bootstrap()

import multiprocessing.spawn
multiprocessing.spawn.get_command_line = get_command_line
multiprocessing.spawn._main = _main

import os
import sys

# Module multiprocessing is organized differently in Python 3.4+
# Python 3.4+
import multiprocessing.popen_spawn_posix as forking

# First define a modified version of Popen.
class _Popen(forking.Popen):
    def __init__(self, *args, **kw):
        if hasattr(sys, 'frozen'):
            # We have to set original _MEIPASS2 value from sys._MEIPASS
            # to get --onefile mode working.
            os.putenv('_MEIPASS2', sys._MEIPASS)
        try:
            super(_Popen, self).__init__(*args, **kw)
        finally:
            if hasattr(sys, 'frozen'):
                # On some platforms (e.g. AIX) 'os.unsetenv()' is not
                # available. In those cases we cannot delete the variable
                # but only set it to the empty string. The bootloader
                # can handle this case.
                if hasattr(os, 'unsetenv'):
                    os.unsetenv('_MEIPASS2')
                else:
                    os.putenv('_MEIPASS2', '')

# Second override 'Popen' class with our modified version.
forking.Popen = _Popen

import sys
from multiprocessing import freeze_support, set_start_method, Process

def test():
    print('In subprocess')

if __name__ == '__main__':
    print(sys.argv,)
    freeze_support()
    set_start_method('spawn')

    print('In main')
    proc = Process(target=test)
    proc.start()
    proc.join()
```

**Comment by xoviat** [comment]
What I am specifically suspecting is that there is an exception raised in `_main`.

**Comment by mkassner** [comment]
Alright! here the output: https://gist.github.com/mkassner/95359c7e2911e9465063c03cb2725b6e

**Comment by xoviat** [comment]
I appear to have a preliminary solution [here][1] / [here][2]. All that remains is to boil it down and then include it in a runtime hook.

[1]: https://travis-ci.org/xoviat/pyinstaller_mac_spawn/builds/202848438
[2]: https://github.com/xoviat/pyinstaller_mac_spawn/blob/master/test.py

**Comment by mkassner** [comment]
@xoviat  I tired your solution briefly but still saw the same failure. Now it appears that your test repo is gone. Any ideas?

**Comment by mkassner** [comment]
I see now that you just renamed the repo. I just ran the script as bundle again and I can see that we go into the subprocess but I m stuck in a loop. So I m seeing endless prints of `in subprocess` and `in main` . Here the log: https://gist.github.com/mkassner/a7a41d5abcd49258fe2c5bcea9324673

**Comment by xoviat** [comment]
> So I m seeing endless prints of in subprocess and in main.

Would you mind providing the spec file?

**Comment by mkassner** [comment]
Im using python3.6 installed via homebrew
this is what I m running on your script (which I named myapp.py)
```shell
pyinstaller myapp.py
./dist/myapp/myapp
```

below the spec file myapp.spec:
```python
# -*- mode: python -*-

block_cipher = None


a = Analysis(['myapp.py'],
             pathex=['/Users/mkassner/Pupil'],
             binaries=[],
             datas=[],
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          exclude_binaries=True,
          name='myapp',
          debug=False,
          strip=False,
          upx=True,
          console=True )
coll = COLLECT(exe,
               a.binaries,
               a.zipfiles,
               a.datas,
               strip=False,
               upx=True,
               name='myapp')
```

**Comment by willpatera** [comment]
@xoviat - any further insight/update into this issue?

**Comment by xoviat** [comment]
@htgoebel A bit of help please. Here is the test that I am using in tests/functional/test_runtime.py:

```python
@skipif_notosx
def test_issue_2322(pyi_builder, capfd):
    pyi_builder.test_source(
        """
        from multiprocessing import set_start_method, Process
        from multiprocessing.spawn import freeze_support
        def test():
            print('In subprocess')
        if __name__ == '__main__':
            freeze_support()
            set_start_method('spawn')
            print('In main')
            proc = Process(target=test)
            proc.start()
            proc.join()
        """)

    out, err = capfd.readouterr()
    assert "In main\nIn subprocess" in out
```

And here is the collection failure:

```=========================== short test summary info ===========================
ERROR tests/functional/test_runtime.py
=================================== ERRORS ====================================
______________ ERROR collecting tests/functional/test_runtime.py ______________
tests\functional\test_runtime.py:13: in <module>
    from pyinstaller.utils.tests import skipif_notosx
pyinstaller.py:15: in <module>
    run()
PyInstaller\__main__.py:70: in run
    args = parser.parse_args(pyi_args)
c:\python36-x64\lib\argparse.py:1733: in parse_args
    self.error(msg % ' '.join(argv))
c:\python36-x64\lib\argparse.py:2389: in error
    self.exit(2, _('%(prog)s: error: %(message)s\n') % args)
c:\python36-x64\lib\argparse.py:2376: in exit
    _sys.exit(status)
E   SystemExit: 2
------------------------------- Captured stderr -------------------------------
usage: py.test [-h] [-v] [-D] [-F] [--specpath DIR] [-n NAME]
               [--add-data <SRC;DEST or SRC:DEST>]
               [--add-binary <SRC;DEST or SRC:DEST>] [-p DIR]
               [--hidden-import MODULENAME] [--additional-hooks-dir HOOKSPATH]
               [--runtime-hook RUNTIME_HOOKS] [--exclude-module EXCLUDES]
               [--key KEY] [-d] [-s] [--noupx] [-c] [-w]
               [-i <FILE.ico or FILE.exe,ID or FILE.icns>]
               [--version-file FILE] [-m <FILE or XML>] [-r RESOURCE]
               [--uac-admin] [--uac-uiaccess] [--win-private-assemblies]
               [--win-no-prefer-redirects]
               [--osx-bundle-identifier BUNDLE_IDENTIFIER] [--distpath DIR]
               [--workpath WORKPATH] [-y] [--upx-dir UPX_DIR] [-a] [--clean]
               [--log-level LEVEL]
               scriptname [scriptname ...]
py.test: error: unrecognized arguments: --maxfail --durations=10 --junitxml=junit-results.xml tests/unit tests/functional -k not tests/functional/test_libraries.py
```

**Comment by htgoebel** [comment]
@xoviat Looks like pytest does not like the arguments (but could be more precise). Did you `pip install -r tests/requirements-tools.txt`? Or maybe the "not …" needs to be quoted.

**Comment by xoviat** [comment]
Failure occurs from [appveyor job](https://ci.appveyor.com/project/xoviat/pyinstaller/build/job/it4h18g12exeejp3) and the [decorator was copied](https://github.com/pyinstaller/pyinstaller/blob/37b4046af5d44a1687041658c989dcea78879d02/tests/functional/test_basic.py#L524).

**Comment by xoviat** [comment]
Update: I'm an idiot. Python is case sensitive.

**Comment by mkassner** [comment]
I m still getting the same error? Any idea why?

**Comment by willpatera** [comment]
@mkassner Unfortunately it looks like tests fail python3.6 see [travisci](https://travis-ci.org/pyinstaller/pyinstaller/builds/211604338)

**Comment by xoviat** [comment]
@willpatera See #2428.

@mkassner Currently waiting on #2508 to be merged, so that I can validate the test.

**Comment by xoviat** [comment]
@mkassner You can see the test that I am using on the associated PR. If you are still having issues, then you will need to submit a new PR with a test case that will fail on travis mac. You can checkout my 'osx' branch to get mac testing enabled, as it will be merged.

**Comment by willpatera** [comment]
@xoviat I have just installed your [osx branch](https://github.com/xoviat/pyinstaller/tree/osx), and tested with following simple `myapp.py` example from @mkassner's [comment above](https://github.com/pyinstaller/pyinstaller/issues/2322#issuecomment-281372210) - with python3.6 installed on macOS `v10.12.3`

```python
import sys
from multiprocessing import freeze_support, set_start_method, Process

def test():
    print('In subprocess')

if __name__ == '__main__':
    print(sys.argv,)
    freeze_support()
    set_start_method('spawn')

    print('In main')
    proc = Process(target=test)
    proc.start()
    proc.join()
```

And created the app with: 

```
pyinstaller myapp.py
```

When trying to execute the app I get the same error as @mkassner [posted](https://github.com/pyinstaller/pyinstaller/issues/2322#issuecomment-278609084)

**Comment by xoviat** [comment]
@willpatera When you run the test, does it fail?

**Comment by willpatera** [comment]
Hi @xoviat I ran the `py.test` from your branch with the following [results](https://gist.github.com/willpatera/36f6c007d27297334d598475ab7bb4a2)

**Comment by xoviat** [comment]
No, the command is:

```
py.test tests/functional/test_regression.py::test_issue_2492 

```

**Comment by willpatera** [comment]
Hi @xoviat I just ran the test above and the test *passed* with the following output:

```
platform darwin -- Python 3.6.0, pytest-3.0.7, py-1.4.33, pluggy-0.4.0 -- /Users/wrp/.virtualenvs/pyinstaller/bin/python3.6
cachedir: .cache
rootdir: /Users/wrp/Desktop/Sandbox/pyinstaller_test/pyinstaller, inifile: setup.cfg
collected 2 items 

tests/functional/test_regression.py::test_issue_2492 PASSED
```

**Comment by xoviat** [comment]
@willpatera I revised my comment shortly after posting it.

**Comment by willpatera** [comment]
@xoviat 

```
platform darwin -- Python 3.6.0, pytest-3.0.7, py-1.4.33, pluggy-0.4.0 -- /Users/wrp/.virtualenvs/pyinstaller/bin/python3.6
cachedir: .cache
rootdir: /Users/wrp/Desktop/Sandbox/pyinstaller_test/pyinstaller, inifile: setup.cfg
collecting 2 items
=================================== no tests ran in 0.01 seconds ===================================
ERROR: not found: /Users/wrp/Desktop/Sandbox/pyinstaller_test/pyinstaller/tests/functional/test_runtime.py::test_issue_2322
(no name '/Users/wrp/Desktop/Sandbox/pyinstaller_test/pyinstaller/tests/functional/test_runtime.py::test_issue_2322' in any of [<Module 'tests/functional/test_runtime.py'>])
```

**Comment by xoviat** [comment]
It's definitely there on my branch.

https://travis-ci.org/pyinstaller/pyinstaller/jobs/213933252#L2718

**Comment by willpatera** [comment]
@xoviat `test_issue_2322` passes on (`rth_mac_spawn` branch)[https://github.com/xoviat/pyinstaller/tree/rth_mac_spawn] (**note** test_issue_2322 does not exist on the osx branch). 

When using the `rth_mac_spawn` branch with the my_app.py example from [above](https://github.com/pyinstaller/pyinstaller/issues/2322#issuecomment-289686130) there is an infinite loop.

**Comment by xoviat** [comment]
@willpatera The test that I am using [appears to be almost identical][1] to the example you gave:

```python
@skipif_notosx
 @xfail_py2
 def test_issue_2322(pyi_builder, capfd):
     pyi_builder.test_source(
         """
         from multiprocessing import set_start_method, Process
         from multiprocessing.spawn import freeze_support
         from multiprocessing.util import log_to_stderr
 
         def test():
             print('In subprocess')
 
         if __name__ == '__main__':
             log_to_stderr()
             freeze_support()
             set_start_method('spawn')
 
             print('In main')
             proc = Process(target=test)
             proc.start()
             proc.join()
         """)
 
     out, err = capfd.readouterr()
 
     # Print the captured output and error so that it will show up in the test output.
     sys.stderr.write(err)
     sys.stdout.write(out)
 
     expected = ["In main", "In subprocess"]
 
     assert "\n".join(expected) in out
     for substring in expected:
         assert out.count(substring) == 1
```

[1]: https://github.com/pyinstaller/pyinstaller/pull/2505/files#diff-84021eb5d294418d7c61f8ae9ac7f4f3R30

**Comment by xoviat** [comment]
> note test_issue_2322 does not exist on the osx branch

That's because the osx branch has nothing to do with this problem. If you look at the pull request, it's about testing on the main repository rather than the pyinstaller_osx_tests repository.

**Comment by xoviat** [comment]
Oh, maybe it's that you have:

```
from multiprocessing import freeze_support, set_start_method, Process
```

And I have:

```
from multiprocessing.spawn import freeze_support
```

I could alias the in the runtime hook

**Comment by willpatera** [comment]
@xoviat - Thanks for this update your import statement works 👍 

```
from multiprocessing.spawn import freeze_support
```

**Comment by xoviat** [comment]
@htgoebel What do you think about overriding the `set_start_method` with a custom function that calls the correct `freeze_support`?

**Comment by htgoebel** [comment]
@xoviat Worth a discussion..

**Comment by xoviat** [comment]
Alright, this should be repaired with the update. The new method is 

```python
         from multiprocessing import set_start_method, Process, freeze_support
 
         def test():
             print('In subprocess')
 
         if __name__ == '__main__':
             freeze_support()
             set_start_method('spawn')
 
             print('In main')
             proc = Process(target=test)
             proc.start()
             proc.join()
```

**Comment by mkassner** [comment]
@xoviat  I just tried your current banch https://github.com/xoviat/pyinstaller/commits/rth_mac_spawn in one of the commits from Apr 5, 2017 there is a regression.

The error im getting in the boot-stap phase is something like this:

```
File "main.py", line 251, in <module>
  File "multiprocessing/spawn.py", line 74, in freeze_support
  File "multiprocessing/spawn.py", line 105, in spawn_main
  File "multiprocessing/spawn.py", line 114, in _main
  File "multiprocessing/spawn.py", line 225, in prepare
  File "multiprocessing/spawn.py", line 277, in _fixup_main_from_path
  File "runpy.py", line 261, in run_path
  File "runpy.py", line 231, in _get_code_from_file
FileNotFoundError: [Errno 2] No such file or directory: '/Users/mkassner/Pupil/pupil_code/deployment/deploy_capture/main.py'
Failed to execute script main
```
When I checkout c586d17f1a87f31d080a15b375bf76345dc29bd0 the error is gone.

**Comment by xoviat** [comment]
Thanks. This is on hold until #2508 is merged.

**Comment by bpteague** [comment]
This is also an issue with spawn mode on Linux (and other POSIX systems, I assume.)  I can verify that #2505 fixes my issue when I modify the runtime hook to run on all POSIX oses.

Additionally - this runtime hook incorporates most (all?) of the strategy used in the [Multiprocessing Recipe](https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Multiprocessing).  I have yet to verify that this also supports spawn mode on Windows; I will do so shortly.

**Comment by bpteague** [comment]
Another data point: this breaks semaphores on Linux, and likely on OSX as well (I will check when I get home and have a Mac to test with.)  The test case in #2505 works fine, but the following throws an error:

```python
from multiprocessing import set_start_method, Process
from multiprocessing import freeze_support
from multiprocessing.util import log_to_stderr
import multiprocessing as mp

def test(s):
    s.acquire()
    print('In subprocess')
    s.release()

if __name__ == '__main__':
    log_to_stderr()
    freeze_support()
    set_start_method('spawn')

    print('In main')
    s = mp.Semaphore()
    s.acquire()
    proc = Process(target=test, args = [s])
    proc.start()
    s.release()
    proc.join()
```

```
brian@tanet:~/src/cytoflow/mp_test$ dist/test_semaphore/test_semaphore 
In main
In subprocess
Traceback (most recent call last):
  File "multiprocessing/util.py", line 252, in _run_finalizers
  File "multiprocessing/util.py", line 185, in __call__
  File "multiprocessing/synchronize.py", line 89, in _cleanup
  File "multiprocessing/semaphore_tracker.py", line 73, in unregister
  File "multiprocessing/semaphore_tracker.py", line 82, in _send
BrokenPipeError: [Errno 32] Broken pipe
```

It seems like the SemaphoreTracker process isn't around, which is where the `BrokenPipeError` comes from.

**Comment by bpteague** [comment]
So I've been poking at this all evening, and it turns out the semaphore tracker process is actually the culprit behind the original issue.

**Background**
When in spawn mode on a POSIX OS (Mac, Linux), Python 3.4+ uses a separate process, the semaphore tracker, to keep track of all the named semaphores that the program creates.  This is because named semaphores are a finite resource and if they aren't freed by the program, they wont' be freed until the next reboot.  So we don't want to leak them.

**Underlying issue**
Unfortunately, Python is stupid about how it creates this background process: it basically spawns another copy of the executable with `-c from multiprocessing.semaphore_tracker import main;main(%d)` on the command line.  (The `%d` is interpolated with the file descriptor of the pipe over which the process receives the names of all of the named semaphores.  The call also includes some other flags, which is what @xoviat's proposed fixed was looking for.)  This is fine when the executable is the Python interpreter, but in a frozen process it creates another copy of the frozen process.

Usually in spawn mode, newly spawned processes have `--multiprocessing-fork` on the command line as well as a handle to a pipe over which the receive the picked `Process` that they are to run.  This magic gets caught by the call to `freeze_support()` which should be at the beginning of `__main__`.  Unfortunately, the code that spawns the semaphore tracker doesn't use this magic.  Thus the infinite loop of new processes.

**Proposed fix**
I can think of two.  The first is to override `freeze_support` in POSIX platforms to look for the particular content of the `-c` flag that we know the spawn code will use to try to start the semaphore tracker, and start it ourselves.

The second is to override `SemaphoreTracker.ensure_running`, which is the function that actually starts the tracker, to use the `--multiprocessing-fork` magic (on POSIX OSes in spawn or forkserver mode).  These approaches may not be 100% mutually exclusive.

I've been going at this all afternoon and evening.  This is a blocker for me too; hopefully I'll get back to it this weekend or next week. @xoviat, you are also more than welcome to have a go.

**Comment by mkassner** [comment]
I can confirm the semaphore tracker issue on MacOS as well.

On Jul 14, 2017 03:49, "Brian Teague" <notifications@github.com> wrote:

So I've been poking at this all evening, and it turns out the semaphore
tracker process is actually the culprit behind the original issue.

*Background*
When in spawn mode on a POSIX OS (Mac, Linux), Python 3.4+ uses a separate
process, the semaphore tracker, to keep track of all the named semaphores
that the program creates. This is because named semaphores are a finite
resource and if they aren't freed by the program, they wont' be freed until
the next reboot. So we don't want to leak them.

*Underlying issue*
Unfortunately, Python is stupid about how it creates this background
process: it basically spawns another copy of the executable with -c from
multiprocessing.semaphore_tracker import main;main(%d) on the command line.
(The %d is interpolated with the file descriptor of the pipe over which the
process receives the names of all of the named semaphores. The call also
includes some other flags, which is what @xoviat <https://github.com/xoviat>'s
proposed fixed was looking for.) This is fine when the executable is the
Python interpreter, but in a frozen process it creates another copy of the
frozen process.

Usually in spawn mode, newly spawned processes have --multiprocessing-fork
on the command line as well as a handle to a pipe over which the receive
the picked Process that they are to run. This magic gets caught by the call
to freeze_support() which should be at the beginning of __main__.
Unfortunately, the code that spawns the semaphore tracker doesn't use this
magic. Thus the infinite loop of new processes.

*Proposed fix*
I can think of two. The first is to override freeze_support in POSIX
platforms to look for the particular content of the -c flag that we know
the spawn code will use to try to start the semaphore tracker, and start it
ourselves.

The second is to override SemaphoreTracker.ensure_running, which is the
function that actually starts the tracker, to use the --multiprocessing-fork
magic (on POSIX OSes in spawn or forkserver mode). These approaches may not
be 100% mutually exclusive.

I've been going at this all afternoon and evening. This is a blocker for me
too; hopefully I'll get back to it this weekend or next week. @xoviat
<https://github.com/xoviat>, you are also more than welcome to have a go.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/pyinstaller/pyinstaller/issues/2322#issuecomment-315247370>,
or mute the thread

**Comment by mkassner** [comment]
@bpteague were you able to devise a solution on these two issues? We use pyinstaller for deployment but Im not familiar enough with the internals to fix the issue.

**Comment by bpteague** [comment]
@mkassner no, I'm still working on it.  With a good handle on the underlying issue, a fix shouldn't be too far away...

**Comment by bpteague** [comment]
(another note: PyInstaller is kind of a bystander here.  the real issue is with the Python multiprocessing library: spawn mode is broken in all frozen executables on POSIX OSes.)

**Comment by bpteague** [comment]
I think I've fixed the hook.  I haven't tested it yet except in the short example above.  I will test it this evening in my larger program on more platforms (mac, windows, linux).

Here's a gist:
https://gist.github.com/bpteague/bc574e4d92c9d9904ceba895fb694fc8

**Comment by bpteague** [comment]
I've updated the gist and confirmed that it works for me on Linux and MacOS.  I have yet to try it with fork and forkserver mode.  YMMV.

**Comment by bpteague** [comment]
I can also confirm that the gist above works for me on Windows. I still haven't tried fork or forkserver mode on POSIX OSes.

**Comment by mkassner** [comment]
I can confirm that this rhook works on both MacOS and Linux. python 3.6.

This saved my so much hassle Thank you @bpteague .
@xoviat  I recommend using this fix!

**Comment by xoviat** [comment]
@bpteague Can you submit a PR?

**Comment by xoviat** [comment]
I will close my PR on this issue as soon as a replacement is submitted.

**Comment by htgoebel** [comment]
#2505, which says it closes this issue, is merged.

**Comment by aprowe** [comment]
I am on same specs as @CKrawczyk, and also run into the same problem. Will try to boil it down to a simple example.

**Comment by pepedelosbosques** [comment]
Same here, sklearn 0.19.2 and 0.20.2

**Comment by pepedelosbosques** [comment]
Just call multiprocessing.freeze_support() at first near the top of your program.

**Comment by kingkastle** [comment]
So, yes, @aprowe your main.pyshould look like this:
```
if __name__ == '__main__':
    from multiprocessing import freeze_support
    freeze_support()
    from multiprocessing import set_start_method, Process
    import my_package  # this package uses `scikit-learn`
    import gooey
   
gui = gooey.Gooey()(my_package.main)
    set_start_method('fork')
    proc = Process(target=gui)
    proc.start()
    proc.join()
```

This will allow your program run. But if you are going to use any multiprocessing, it is going to fail. :(

**Comment by htgoebel** [comment]
@kingkastle To be on the safe side the call to `freeze_support()`  must be done prior to all  other imports. Move it as far to teh top as possible.

**Comment by kingkastle** [comment]
Hey @htgoebel according to [python docs](https://docs.python.org/3.7/library/multiprocessing.html#multiprocessing.freeze_support) it should be the first thing after ``if __name__ == '__main__':`` which I think is what I have done.

**Comment by htgoebel** [comment]
@kingkastle This docs clearly say " For example:".

**Comment by CliffMitchell** [comment]
I hope this thread is still active because I have the same problem.

I have a Python3/PyQT5 app, frozen using PyInstaller. On macOS my Python3 code runs fine from a CLI. The version frozen using PyInstaller also runs successfully but creates new copies of itself every couple of seconds in what looks like an infinite loop. 

By way of background I have successfully frozen this app on Ubuntu 18.05 and Windows 10 without seeing this problem.

I am running Python 3.7.2 with PyInstaller 3.4 and PyQT5 5.12, on macOS Mojave 10.14.3.

Given the Python code runs perfectly from the CLI and from PyCharm, and the frozen package runs (but with multiple copies of itself) it looks like the problem is somewhere in the freezing process. I am not directly using multiprocessing but am using scikit-learn that does, I believe, use multiprocessing. I have tried calling freeze_support() at the top, and various other places but none fix the problem. 

Like CKrawczyk my .spec file does have sklearn as a hidden import. I know that freeze_support() should be called before sklearn so I wonder if having sklearn as a hidden import in the .spec file is somehow causing the problem? Other than that I have hit a brick wall. Can anyone suggest how we can get to the bottom of this problem please?

**Comment by Serlopal** [comment]
I'm in the same boat as @CliffMitchell. My PyQt5 app (I use this unofficial conversion for 2.7 https://github.com/pyqt/python-qt5 as the core library of the roject, platformio, does not support python 3+) runs into an infinite loop and just keeps swpaning new copies of the main window forever.

I have checked exerimentally that the cause of the issue is a call to Subprocess.call inside the platformio library. I have of course used the infamous line "subprocess.freeze_support" without any success. I have also gone through the Pyinstaller "recipies" in the repo for subprocess and multiprocessing libraries.

At this point, I really don't know what else to do. The app works perfectly fine in PyCharm and this problem only arises once called from the .exe generated by Pyinstaller (happens also with --onedir option outputting a folder version of the program)

Any help would be appreciated.

Many thanks

**Comment by eclairank** [comment]
i solved!

> https://github.com/mherrmann/fbs/issues/87#issuecomment-471489980

**Comment by JapinderSandhu** [comment]
> i solved!
> 
> > [mherrmann/fbs#87 (comment)](https://github.com/mherrmann/fbs/issues/87#issuecomment-471489980)

Thanks ! This actually worked for me: 

Before:

    multiprocessing.set_start_method('spawn')

After:

    multiprocessing.freeze_support()
    multiprocessing.set_start_method('spawn')
